public/src/data/save.js

```diff
@@ -0,0 +1,333 @@
+import { scopeList } from '../circuit';
+import { resetup } from '../setup';
+import { update } from '../engine';
+import { stripTags, showMessage } from '../utils';
+import { backUp } from './backupCircuit';
+import simulationArea from '../simulationArea';
+import backgroundArea from '../backgroundArea';
+import { findDimensions } from '../canvasApi';
+import { projectSavedSet } from './project';
+
+/**
+ * Function to set the name of project.
+ * @param {string} name - name for project
+ * @category data
+ */
+export function setProjectName(name) {
+    name = stripTags(name);
+    var projectName = name;
+    $('#projectName').html(name);
+}
+
+/**
+ * Helper function to save canvas as image based on image type
+ * @param {string} name -name of the circuit
+ * @param {string} imgType - image type ex: png,jpg etc.
+ * @category data
+ */
+function downloadAsImg(name, imgType) {
+    const gh = simulationArea.canvas.toDataURL(`image/${imgType}`);
+    const anchor = document.createElement('a');
+    anchor.href = gh;
+    anchor.download = `${name}.${imgType}`;
+    anchor.click();
+}
+
+/**
+ * Generates JSON of the entire project
+ * @param {string} name - the name of project
+ * @return {JSON}
+ * @category data
+ */
+export function generateSaveData(name) {
+    data = {};
+
+    // Prompts for name, defaults to Untitled
+    name = projectName || name || prompt('Enter Project Name:') || 'Untitled';
+    data.name = stripTags(name);
+    projectName = data.name;
+    setProjectName(projectName);
+
+    // Save project details
+    data.timePeriod = simulationArea.timePeriod;
+    data.clockEnabled = simulationArea.clockEnabled;
+    data.projectId = projectId;
+    data.focussedCircuit = globalScope.id;
+
+    // Project Circuits, each scope is one circuit
+    data.scopes = [];
+    const dependencyList = {};
+    const completed = {};
+
+    // Getting list of dependencies for each circuit
+    for (id in scopeList) { dependencyList[id] = scopeList[id].getDependencies(); }
+
+    // Helper function to save Scope
+    // Recursively saves inner subcircuits first, before saving parent circuits
+    function saveScope(id) {
+        if (completed[id]) return;
+
+        for (let i = 0; i < dependencyList[id].length; i++) {
+            // Save inner subcircuits
+            saveScope(dependencyList[id][i]);
+        }
+
+        completed[id] = true;
+
+        update(scopeList[id], true); // For any pending integrity checks on subcircuits
+
+        data.scopes.push(backUp(scopeList[id]));
+    }
+
+    // Save all circuits
+    for (id in scopeList) { saveScope(id); }
```

Comment:
Don't use 'in' for loop. Reason: [check this](https://stackoverflow.com/questions/500504/why-is-using-for-in-with-array-iteration-a-bad-idea)